pkg/reconciler/v1alpha1/clusteringress/resources/virtual_service.go

```diff
@@ -0,0 +1,158 @@
+	}
+	vs.Labels[networking.IngressLabelKey] = ci.Name
+
+	ingressLabels := ci.Labels
+	vs.Labels[serving.RouteLabelKey] = ingressLabels[serving.RouteLabelKey]
+
+	return vs
+}
+
+func makeVirtualServiceSpec(ci *v1alpha1.ClusterIngress) *v1alpha3.VirtualServiceSpec {
+	spec := v1alpha3.VirtualServiceSpec{
+		// We want to connect to two Gateways: the Knative shared
+		// Gateway, and the 'mesh' Gateway.  The former provides
+		// access from outside of the cluster, and the latter provides
+		// access for services from inside the cluster.
+		Gateways: []string{
+			names.K8sGatewayFullname,
+			"mesh",
+		},
+		Hosts: getHosts(ci),
+	}
+
+	for _, rule := range ci.Spec.Rules {
+		hosts := rule.Hosts
+		for _, p := range rule.HTTP.Paths {
+			spec.Http = append(spec.Http, *makeVirtualServiceRoute(hosts, &p))
+		}
+	}
+	return &spec
+}
+
+func makePortSelector(ios intstr.IntOrString) v1alpha3.PortSelector {
+	if ios.Type == intstr.Int {
+		return v1alpha3.PortSelector{
+			Number: uint32(ios.IntValue()),
+		}
+	}
+	return v1alpha3.PortSelector{
+		Name: ios.String(),
+	}
+}
+
+func makeVirtualServiceRoute(hosts []string, http *v1alpha1.HTTPClusterIngressPath) *v1alpha3.HTTPRoute {
+	matches := []v1alpha3.HTTPMatchRequest{}
+	for _, host := range hosts {
+		matches = append(matches, makeMatch(host, http.Path))
+	}
+	weights := []v1alpha3.DestinationWeight{}
+	for _, split := range http.Splits {
+		weights = append(weights, v1alpha3.DestinationWeight{
+			Destination: v1alpha3.Destination{
+				Host: reconciler.GetK8sServiceFullname(
+					split.Backend.ServiceName, split.Backend.ServiceNamespace),
+				Port: makePortSelector(split.Backend.ServicePort),
+			},
+			Weight: split.Percent,
+		})
+	}
+	return &v1alpha3.HTTPRoute{
+		Match:   matches,
+		Route:   weights,
+		Timeout: timeoutToString(http.Timeout.Duration),
+		Retries: &v1alpha3.HTTPRetry{
+			Attempts:      http.Retries.Attempts,
+			PerTryTimeout: timeoutToString(http.Retries.PerTryTimeout.Duration),
+		},
+		AppendHeaders: http.AppendHeaders,
+	}
+}
+
+func makeMatch(host string, pathRegExp string) v1alpha3.HTTPMatchRequest {
+	match := v1alpha3.HTTPMatchRequest{
+		Authority: &istiov1alpha1.StringMatch{
+			Exact: host,
+		},
+	}
+	// Empty pathRegExp is considered match all path. We only need to
+	// consider pathRegExp when it's non-empty.
+	if pathRegExp != "" {
+		match.Uri = &istiov1alpha1.StringMatch{
+			Regex: pathRegExp,
+		}
+	}
+	return match
+}
+
+func getHosts(ci *v1alpha1.ClusterIngress) []string {
+	hosts := make(map[string]interface{})
+	unique := []string{}
+	for _, rule := range ci.Spec.Rules {
+		for _, h := range rule.Hosts {
+			if _, existed := hosts[h]; !existed {
+				hosts[h] = true
+				unique = append(unique, h)
+			}
+		}
+	}
+	// Sort the names to give a deterministic ordering.
+	sort.Strings(unique)
+	return unique
+}
+
+func timeoutToString(timeout time.Duration) string {
+	return fmt.Sprintf("%ds", int(timeout.Truncate(time.Second).Seconds()))
```

Comment:
I am a bit worried about using `Truncate` with Second here. For example, if the original timeout is 10.5 seconds, it will be truncated as 10 seconds, which seems not accurate enough.
I am wondering if Istio accepts `milliseconds` as timeout unit. If it does, maybe we can consider to truncate the timeout with milliseconds.

Reply:
Hmm, can we just serialize the `time.Duration`?  `timeout.String()` should do it.

Reply:
The type of the field in istio is indeed of protobuf.Duration, which is made up of `Seconds` and `Nanos`:
```
type Duration struct {
	// Signed seconds of the span of time. Must be from -315,576,000,000
	// to +315,576,000,000 inclusive. Note: these bounds are computed from:
	// 60 sec/min * 60 min/hr * 24 hr/day * 365.25 days/year * 10000 years
	Seconds int64 `protobuf:"varint,1,opt,name=seconds,proto3" json:"seconds,omitempty"`
	// Signed fractions of a second at nanosecond resolution of the span
	// of time. Durations less than one second are represented with a 0
	// `seconds` field and a positive or negative `nanos` field. For durations
	// of one second or more, a non-zero value for the `nanos` field must be
	// of the same sign as the `seconds` field. Must be from -999,999,999
	// to +999,999,999 inclusive.
	Nanos                int32    `protobuf:"varint,2,opt,name=nanos,proto3" json:"nanos,omitempty"`
	XXX_NoUnkeyedLiteral struct{} `json:"-"`
	XXX_unrecognized     []byte   `json:"-"`
	XXX_sizecache        int32    `json:"-"`
}
```

Have not looked in to that deep, bu I'm afraid it can not understand time duration formatted string so "smartly", and to support milliseconds in timeout, we might need to make it a float num for seconds, e.g "1.00123s"

Reply:
Aha, I underestimated the compatibility of protobuf, it makes use of built-in time functions when unmarshal: https://github.com/gogo/protobuf/blob/master/jsonpb/jsonpb.go#L870,
so it can recognize time duration formatted string. Let's keep the passed in duration as is, and remove the extra truncation.

Reply:
done